V3: fetchWrapper and cacheWrapper

gauntface opened the pull request:
R: @jeffposnick @addyosmani @philipwalton 

This is one possible adaption of the current requestWrapper stuff we have in workbox-runtime-caching at the moment.

# Differences

There are a few major differences between this PR vs current requestWrapper. Below is list of differences and explanations:

- Dropped waitOnCache and cacheKey
- Seperated requestWrapper into fetchWrapper and cacheWrapper.
    - These two changes were hand in hand. waitOnCache and cacheKey were only used by precaching and removing them simplified the overall code for the *Wrapper modules. Seperation was needed to enable precaching to perform these features.
- Removed the restriction of single plugins to manipulate values.
    - Reading through the current requestWrapper code I wasn't sure why some plugin methods could only have callback vs others. I've added some tests to show how multiple plugins will play together and didn't find anything too concerning in terms of a developer understanding. But I may be missing scenarios.
- fetchWrapper and cacheWrapper export static functions
- Plugins are looped through and relevant functions are called if they exist. Previously a map was generated in the RequestWrapper constructor. Will be inefficient if a large number of plugins were used.

# Discussion Points

- API doesn't allow for a global set of plugins at the moment, but we could add a method to workbox-core allowing a developer to set plugins which could be called before or after the specific function plugins.

```javascript
import core from 'workbox-core';
core.setGlobalPlugins([
    ......
]);
```

I still need to account for where cache names come from where they are defined, but again the PR was getting pretty big, so stopped from progressing further. Will work on this in the next PR.

addyosmani commented:
> API doesn't allow for a global set of plugins at the moment, but we could add a method to workbox-core allowing a developer to set plugins which could be called before or after the specific function plugins.

It would be interesting to think about how we're landing on the sketch for the plugins model described in the issue. Do we need to have a separate discussion around how a plugin model here should work?

jeffposnick commented on test/workbox-core/node/utils/test-cacheWrapper.mjs:
```diff
@@ -0,0 +1,159 @@
+        secondPlugin,
+      ]);
+
+      expect(spyOne.callCount).to.equal(1);
```
FWIW, `sinon` has some assertion helpers that would eliminate most of this repetition, i.e. `sinon.assert.calledWithExactly(spy, arg1, arg2, ...);`

http://sinonjs.org/releases/v2.4.1/assertions/

I could understand not waiting to mix `sinon` and `chai` assertions, but... that's a lot of repeated code, as it stands.

gauntface commented on packages/workbox-core/utils/cacheWrapper.mjs:
```diff
@@ -0,0 +1,83 @@
+
+  // Only conditionally await the caching completion, giving developers the
+  // option of returning early for, e.g., read-through-caching scenarios.
+  if (waitOnCache) {
```
cacheWrapper.put will always wait on the operation. The main use for it was so that precaching could block on the cache put when fetchAndCache originally didn't.

With these being seperated, it's up to the implementation to block or not on the operation.

gauntface commented on packages/workbox-core/utils/cacheWrapper.mjs:
```diff
@@ -0,0 +1,83 @@
+};
+
+export default {
+  put: putWrapper,
```
At the moment it hasn't been needed output of put. When it is needd, someone can just expose it.

jeffposnick commented on packages/workbox-core/utils/fetchWrapper.mjs:
```diff
@@ -0,0 +1,55 @@
+  } catch (err) {
+    for (let plugin of failedFetchPlugins) {
+      if (plugin.fetchDidFail) {
+        await plugin.fetchDidFail(
```
The earlier implementation `fetchDidFail` just passes in a clone of the original request and doesn't pass in a (possibly identical) clone of the "filtered" request. I'm okay with passing in the filtered copy, but it would be cleaner if you do that with the overall change to use an object with property names rather than positional parameters to callbacks, just to clarify what each parameter is.

gauntface commented:
@jeffposnick I did a pass of the code and added a  tonne of your feedback (And switched to objects for all the callbacks).

There was one question you raised regarding originalRequest and fetchedRequest in the fetchWrapper. The thinking there is that at the moment the callback is given the fetch request that failed, but the request could have been altered by a plugin, meaning the input that would probably be useful to identify the Route, would be seperated / lost from the request that ultimately failed.

I renamed the arguments to `originalRequest` and `request` so anyone using the plugin today will have the same object, but the original request will get to rename.

jeffposnick commented:
Re: `fetchDidFail`, that sounds fine now that there are named parameters for each of those two requests.

gauntface commented on test/mocks/mock-fetch.js:
```diff
@@ -0,0 +1,8 @@
+global.fetch = async (request) => {
```
For future me: I tried exporting this as a function and then stubbing it out, but that prevents other tests from spying / stubbing it and it didn't seem to work unless global.fetch was something.
